examples/advanced/quantum_volume.py

```diff
@@ -151,19 +247,36 @@ def main(num_qubits: int, depth: int, num_repetitions: int, seed: int):
         num_qubits: The number of qubits for the circuit.
         depth: The number of gate layers to generate.
         num_repetitions: The number of times to run the algorithm.
+        samplers: The samplers to run the algorithm on.
+        deconstruct: An optional function to deconstruct the model circuit's
+            gates down to the targer devices gate set.
+        optimize: An optional function to optimize the circuit's gates
+            after routing.
     """
-    for _ in range(num_repetitions):
+    result = QuantumVolumeResult()
+    for repetition in range(num_repetitions):
+        print(f"Repetition {repetition + 1}")
         model_circuit = generate_model_circuit(
             num_qubits, depth, random_state=np.random.RandomState(seed))
+        result.model_circuits.append(model_circuit)
+
         heavy_set = compute_heavy_set(model_circuit)
-        print(f"Heavy Set: {heavy_set}")
-        print(f"Ideal simulation probability: "
-              f"{sample_heavy_set(model_circuit, heavy_set)}")
-        noisy = cirq.DensityMatrixSimulator(noise=cirq.ConstantQubitNoiseModel(
-            qubit_noise_gate=cirq.DepolarizingChannel(p=0.005)))
-        print(f"Noisy simulation probability: "
-              f"{sample_heavy_set(model_circuit, heavy_set, sampler=noisy)}")
-        # TODO(villela): Implement model circuit and run it.
+        result.heavy_sets.append(heavy_set)
+        print(f"  Heavy Set: {heavy_set}")
+
+        [compiled_circuit, mapping] = compile_circuit(model_circuit,
+                                                      device=device,
+                                                      deconstruct=deconstruct,
+                                                      optimize=optimize)
+        result.compiled_circuits.append(compiled_circuit)
+        for idx, sampler in enumerate(samplers):
+            prob = sample_heavy_set(compiled_circuit,
```

Comment:
outside the scope of this PR, but:

This is a weird loop to have because it indicates that your functions aren't very composable. Prefer separating your script into two "tasks": 1. problem generation 2. problem execution. You can imagine wanting to do all the hard hog computation up front (once) and then collect data using various samplers on a whim, perhaps at regular time intervals. Imagine we'd want to run this benchmark once a week to track performance of a device over time

Reply:
Ah sure, that makes sense. Is it OK if I do that in a PR right after this one, so this doesn't get too big?

Reply:
of course